### ISSUE TITLE
README: misleading information regarding go-set-project for projectile.el

### ISSUE THREAD

**Issue by db47h** [dominikh/go-mode.el#174]
The readme says:

> - GOPATH detection – the function go-guess-gopath will guess a suitable value for GOPATH (...)
>
>   You can either call go-set-project manually, or integrate it with Projectile's project switching hooks, or any other means of switching projects you may employ.

In fact, when the user manually opens a file (via `find-file` or any wrapper around it), projectile does not call any of the `projectile-<before|after>-switch-project-hook` hooks. These are called only when using  `projectile-switch-project` interactively. However projectile hooks into `find-file-hook` to switch project on the fly.

**TL;DR:** ~~For projectile users, to make sure that `go-set-project` is always set properly, it needs to be called from `go-mode-hook` itself.~~

**Comment by db47h** [comment]
This also applies to https://github.com/dominikh/go-mode.el/blob/master/go-mode.el#L1949

**Comment by db47h** [comment]
After some more careful thinking about it, there are three correlated issues here:

1. Calling `go-set-project` for every opened Go file is probably not correct: what would/should happen when opening a file from the standard library via `godef-jump`
2. GOPATH is updated globally: when switching buffers, it will still be set to the GOPATH of the last opened file. A workaround for this point would be to have a buffer local variable, say `go-path`, and wrap every shell command in a construct like this:

```emacs-lisp
(let ((process-environment (cons (concat "GOPATH=" go-path) process-environment)))
    (compile ...)
)
```

3. gocode needs to be restarted every time GOPATH changes.

So my initial suggestion of hooking `go-set-project` into `go-mode-hook` is plain wrong unless these issues are addressed. Any suggestions?

**Comment by dominikh** [comment]
You have essentially discovered why we suggest hooking into Projectile's project switching hooks, and don't react to files being opened or buffers being switched. I will now expand why this is the least bad solution and why any means of automatic setting of GOPATH are futile.

The main issue is that buffer-local variables will not be inherited by "related" buffers. This makes a buffer-local process environment useless because a lot of commands will operate on their own buffer and use the global default. One could, as you have pointed out, wrap all of these commands, in a `let`, but that is not feasible. We do not control all Emacs packages and the majority of them doesn't care or know about go-mode, yet might operate on Go buffers and may need to use the correct GOPATH. 

One could hook into buffer switching hooks and set the global GOPATH, but that still partially suffers from the same issue as buffer local variables, as not all operations on buffers necessarily switch to these buffers first. Additionally, Emacs switches buffers internally several times a second during normal operation, so hooking into this would have a negative effect on performance.

This is precisely why we suggest using Projectile, or a similar tool, and to consciously switch between projects, as Emacs itself lacks a concept of the same. This is in line with other programming environments in Emacs, such as the Python one. Specifically, when going from project A to B, you wouldn't use switch-buffer, but the Projectile equivalent, to signal that you're working in the other project now.

As for gocode: gocode needn't be restarted. The gocode client will pass its environment to the gocode server, which in turn acts on the correct GOPATH. Of course this, again, requires the GOPATH to be set globally, not locally to the buffer.

tl;dr: You've run into a limitation in Emacs itself, and switching between projects needs to be explicit.

**Comment by db47h** [comment]
Alright, so projectile it is (or any of its like) and explicit project switching. I don't have any issue with this. However, the few pointers in the README still don't work as expected. Using projectile's project switching functions works, but they only offer to switch to already known projects. When one wants to work on a new project, and make it known to projectile, it's just a matter of `find-file` one of the project files, but in that case, no project switching hook is called. Same thing happens when starting emacs from the command line with a file (in a known project) as argument: no project switching hooks are called.

I'm fully aware that this is a limitation of projectile, for more or less the same reasons that you mentioned above. Still, a bit more details in the README could help new users, and may be prevent them from rediscovering all this the hard way ;)

**Comment by dominikh** [comment]
GOPATH has been replaced by Go modules, making go-set-project unnecessary and obsolete. We no longer mention it in the README.
